pkg/printer/printer.go

```diff
@@ -2,31 +2,75 @@ package printer
 
 import (
 	"fmt"
+	"os"
 
 	"github.com/doitintl/kube-no-trouble/pkg/judge"
 )
 
-var printers = map[string]func() Printer{
+var printers = map[string]func(string) (Printer, error){
 	"json": newJSONPrinter,
 	"text": newTextPrinter,
 }
 
 type Printer interface {
 	Print([]judge.Result) error
+	Close() error
 }
 
-func NewPrinter(choice string) (Printer, error) {
+type commonPrinter struct {
+	outputFile *os.File
+}
+
+// newCommonPrinter creates new printer that prints to given output file
+func newCommonPrinter(outputFileName string) (*commonPrinter, error) {
+	outputFile, err := ensureOutputFile(outputFileName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to ensure output device: %v", err)
+	}
+
+	return &commonPrinter{
+		outputFile: outputFile,
+	}, nil
+}
+
+// NewPrinter creates new printer of given type that prints to given output file
+func NewPrinter(choice string, outputFileName string) (Printer, error) {
 	printer, err := ParsePrinter(choice)
 	if err != nil {
 		return nil, err
 	}
-	return printer(), nil
+	return printer(outputFileName)
+}
+
+// Close will free resources used by the printer
+func (c *commonPrinter) Close() error {
+	if c.outputFile.Name() != os.Stdout.Name() {
+		if err := c.outputFile.Close(); err != nil {
+			return fmt.Errorf("failed to close output file: %w", err)
+		}
+	}
+
+	return nil
 }
 
-func ParsePrinter(choice string) (func() Printer, error) {
+func ParsePrinter(choice string) (func(string) (Printer, error), error) {
 	printer, exists := printers[choice]
 	if !exists {
 		return nil, fmt.Errorf("unknown printer type %s", choice)
 	}
 	return printer, nil
 }
+
+// ensureOutputFile will open file for writing, or create one if non-existent
+func ensureOutputFile(fileName string) (*os.File, error) {
+	if fileName == "-" {
+		return os.Stdout, nil
+	}
+
+	of, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
```

Comment:
does this work for windows?

Reply:
might want to use ModeAppend instead

Reply:
This is commonly used in examples. I'm not sure though, haven't tested on Windows. Why shouldn't it?

Reply:
0644 is a unix term

Reply:
I assume it should work just fine - https://pkg.go.dev/os#FileMode:
> A FileMode represents a file's mode and permission bits. The bits have the same definition on all systems, so that information about files can be moved from one system to another portably. Not all bits apply to all systems. ...

Reply:
That btw. brings up a very good question - we do build on many platforms, incl. Windows. Should we run tests on all of those as well?

I would say yes, but not sure how difficult is the implementation.